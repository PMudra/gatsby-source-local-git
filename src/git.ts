import git, { DiffResult } from "simple-git/promise"
import { DefaultLogFields, ListLogLine } from "simple-git"

interface Author {
  id: string
  name: string
  email: string
}

interface Commit extends ListLogLine {
  id: string
  hash: string
  message: string
  body: string
  refs: string
  date: Date
  latest: boolean
  author: Author
}

interface Tag {
  id: string
  name: string
  latest: boolean
}

interface Branch {
  id: string
  name: string
  commit: string
  current: boolean
}

const createAuthor = (name: string, email: string): Author => ({
  name,
  email,
  id: `${name}__${email}`,
})

const createDiff = (diff: DiffResult): DiffResult => ({
  ...diff,
})

const getCommits = async (): Promise<Commit[]> => {
  const {
    all,
    latest: { hash: latestHash },
  } = await git().log<DefaultLogFields>({ "--stat": 4096 })

  return all.map(
    ({ hash, diff, date, author_email, author_name, ...rest }) => ({
      hash,
      ...rest,
      id: hash,
      date: new Date(date),
      latest: hash === latestHash,
      diff: diff && createDiff(diff),
      author: createAuthor(author_name, author_email),
    })
  )
}

const getTags = async (): Promise<Tag[]> => {
  const { all, latest } = await git().tags()

  return all.map((name) => ({
    name,
    id: name,
    latest: name === latest,
  }))
}

const getBranches = async (): Promise<Branch[]> => {
  const { branches } = await git().branch(["-v", "-a"])

  return Object.values(branches).map(({ name, commit, current }) => ({
    name,
    commit,
    current,
    id: name,
  }))
}

export { getCommits, getTags, getBranches, Author, Commit, Tag, Branch }
