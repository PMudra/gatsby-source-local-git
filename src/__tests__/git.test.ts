import { getBranches, getCommits, getTags } from "../git"

test("First commit should match snapshot", async () => {
  const commits = await getCommits()
  expect(commits.pop()).toMatchSnapshot()
})

test("Exactly one commit should be latest", async () => {
  const commits = await getCommits()
  expect(commits.filter((commit) => commit.latest)).toHaveLength(1)
})

test("First tag should match snapshot", async () => {
  const [firstTag] = await getTags()
  expect(firstTag).toMatchSnapshot()
})

test("Exactly one tag should be latest", async () => {
  const tags = await getTags()
  expect(tags.filter((tag) => tag.latest)).toHaveLength(1)
})

test("Exactly one branch should be current", async () => {
  const branches = await getBranches()
  expect(branches.filter((branch) => branch.current)).toHaveLength(1)
})
