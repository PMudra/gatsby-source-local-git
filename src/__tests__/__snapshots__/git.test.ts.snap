// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First commit should match snapshot 1`] = `
Object {
  "author": Object {
    "email": "philipp.mudra@web.de",
    "id": "Philipp Mudra__philipp.mudra@web.de",
    "name": "Philipp Mudra",
  },
  "body": "",
  "date": 2019-11-10T18:07:04.000Z,
  "diff": Object {
    "changed": 11,
    "deletions": 0,
    "files": Array [
      Object {
        "binary": false,
        "changes": 53,
        "deletions": 0,
        "file": ".devcontainer/Dockerfile",
        "insertions": 1,
      },
      Object {
        "binary": false,
        "changes": 26,
        "deletions": 0,
        "file": ".devcontainer/devcontainer.json",
        "insertions": 1,
      },
      Object {
        "binary": false,
        "changes": 99,
        "deletions": 0,
        "file": ".gitignore",
        "insertions": 1,
      },
      Object {
        "binary": false,
        "changes": 7,
        "deletions": 0,
        "file": ".prettierrc",
        "insertions": 1,
      },
      Object {
        "binary": false,
        "changes": 2,
        "deletions": 0,
        "file": "gatsby-node.js",
        "insertions": 1,
      },
      Object {
        "binary": false,
        "changes": 1,
        "deletions": 0,
        "file": "index.js",
        "insertions": 1,
      },
      Object {
        "binary": false,
        "changes": 15419,
        "deletions": 0,
        "file": "package-lock.json",
        "insertions": 38,
      },
      Object {
        "binary": false,
        "changes": 25,
        "deletions": 0,
        "file": "package.json",
        "insertions": 1,
      },
      Object {
        "binary": false,
        "changes": 56,
        "deletions": 0,
        "file": "src/gatsby-node.ts",
        "insertions": 1,
      },
      Object {
        "binary": false,
        "changes": 25,
        "deletions": 0,
        "file": "src/git.ts",
        "insertions": 1,
      },
      Object {
        "binary": false,
        "changes": 67,
        "deletions": 0,
        "file": "tsconfig.json",
        "insertions": 1,
      },
    ],
    "insertions": 15780,
  },
  "hash": "c003b915fa82ae86adcebd9f9461bbd215fd52eb",
  "id": "c003b915fa82ae86adcebd9f9461bbd215fd52eb",
  "latest": false,
  "message": "Initial commit",
  "refs": "",
}
`;

exports[`First tag should match snapshot 1`] = `
Object {
  "id": "v0.1.2",
  "latest": false,
  "name": "v0.1.2",
}
`;
